File: sdk/java/src/main/java/io/cordum/cap/Testing.java
package io.cordum.cap;

import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.JobRequest;
import io.cordum.cap.agent.v1.JobResult;
import io.cordum.cap.agent.v1.JobStatus;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

public final class Testing {

    private Testing() {}

    /** Tracks published messages for assertions. */
    public static class Published {
        public final String subject;
        public final byte[] data;

        public Published(String subject, byte[] data) {
            this.subject = subject;
            this.data = data;
        }
    }

    /** In-memory mock for NATS-like publish/subscribe for unit tests. */
    public static class MockBus {
        private final CopyOnWriteArrayList<Published> published = new CopyOnWriteArrayList<>();
        private final ConcurrentHashMap<String, List<Consumer<byte[]>>> subscribers = new ConcurrentHashMap<>();

        public void publish(String subject, byte[] data) {
            published.add(new Published(subject, data));
            List<Consumer<byte[]>> handlers = subscribers.get(subject);
            if (handlers != null) {
                for (Consumer<byte[]> h : handlers) {
                    h.accept(data);
                }
            }
        }

        public void subscribe(String subject, Consumer<byte[]> handler) {
            subscribers.computeIfAbsent(subject, k -> new CopyOnWriteArrayList<>()).add(handler);
        }

        public List<Published> published() {
            return new ArrayList<>(published);
        }

        public List<Published> publishedTo(String subject) {
            List<Published> result = new ArrayList<>();
            for (Published p : published) {
                if (p.subject.equals(subject)) {
                    result.add(p);
                }
            }
            return result;
        }

        public void clear() {
            published.clear();
            subscribers.clear();
        }
    }

    /**
     * Submit a job request to a handler and return the result.
     * Useful for quick unit tests without NATS infrastructure.
     */
    public static JobResult submitAndWait(Middleware.Handler handler, JobRequest req) throws Exception {
        Middleware.Context ctx = new Middleware.Context("test-trace", "test-sender", null);
        return handler.handle(ctx, req);
    }

    /** Recording implementation of Metrics for test assertions. */
    public static class RecordingMetrics implements Metrics {
        public final List<String> receivedJobs = new CopyOnWriteArrayList<>();
        public final List<String> completedJobs = new CopyOnWriteArrayList<>();
        public final List<String> failedJobs = new CopyOnWriteArrayList<>();
        public final List<String> heartbeats = new CopyOnWriteArrayList<>();
        public final List<String> errors = new CopyOnWriteArrayList<>();
        public final List<String> progress = new CopyOnWriteArrayList<>();

        @Override public void onJobReceived(String jobId, String topic) { receivedJobs.add(jobId); }
        @Override public void onJobCompleted(String jobId, long durationMs, String status) { completedJobs.add(jobId); }
        @Override public void onJobFailed(String jobId, String errorMsg) { failedJobs.add(jobId); }
        @Override public void onHeartbeatSent(String workerId) { heartbeats.add(workerId); }
        @Override public void onProgressEmitted(String jobId, int percent) { progress.add(jobId); }
        @Override public void onError(String context, String errorMsg) { errors.add(context + ": " + errorMsg); }
    }
}

File: sdk/dotnet/src/Cordum.Cap/Testing.cs
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Tracks published messages for test assertions.
/// </summary>
public sealed class MockBus
{
    private readonly List<(string Subject, byte[] Data)> _published = new();
    private readonly object _lock = new();

    public void Publish(string subject, byte[] data)
    {
        lock (_lock)
        {
            _published.Add((subject, data));
        }
    }

    public int PublishedCount
    {
        get { lock (_lock) { return _published.Count; } }
    }

    public List<byte[]> PublishedTo(string subject)
    {
        lock (_lock)
        {
            return _published
                .Where(p => p.Subject == subject)
                .Select(p => p.Data)
                .ToList();
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            _published.Clear();
        }
    }
}

/// <summary>
/// Submit a job request to a handler and return the result.
/// Useful for quick unit tests without NATS infrastructure.
/// </summary>
public static class TestHelper
{
    public static JobResult SubmitAndWait(HandlerDelegate handler, JobRequest req)
    {
        var ctx = new Context
        {
            TraceId = "test-trace",
            SenderId = "test-sender",
            Packet = null,
        };
        return handler(ctx, req);
    }
}

/// <summary>
/// Recording implementation of IMetricsHook for test assertions.
/// </summary>
public sealed class RecordingMetrics : IMetricsHook
{
    private readonly object _lock = new();

    public List<string> ReceivedJobs { get; } = new();
    public List<string> CompletedJobs { get; } = new();
    public List<string> FailedJobs { get; } = new();
    public List<string> Heartbeats { get; } = new();
    public List<string> ProgressEvents { get; } = new();
    public List<string> Errors { get; } = new();

    public void OnJobReceived(string jobId, string topic)
    {
        lock (_lock) { ReceivedJobs.Add(jobId); }
    }

    public void OnJobCompleted(string jobId, long durationMs, string status)
    {
        lock (_lock) { CompletedJobs.Add(jobId); }
    }

    public void OnJobFailed(string jobId, string errorMsg)
    {
        lock (_lock) { FailedJobs.Add(jobId); }
    }

    public void OnHeartbeatSent(string workerId)
    {
        lock (_lock) { Heartbeats.Add(workerId); }
    }

    public void OnProgressEmitted(string jobId, int percent)
    {
        lock (_lock) { ProgressEvents.Add(jobId); }
    }

    public void OnError(string context, string errorMsg)
    {
        lock (_lock) { Errors.Add($"{context}: {errorMsg}"); }
    }
}
